File: aquarium/aquarium.py
import asyncio
from machine import Pin, PWM  # type: ignore
from communication_protocol.communication_module import CommunicationModule
from communication_protocol.communication_protocol import DeviceMessage
from communication_protocol.message import accept_message


class Aquarium:
    def __init__(
        self,
        pin_r: int,  # D5
        pin_g: int,  # D6
        pin_b: int,  # D7
        pin_fluo: int,  # D3
        communication_protocol: CommunicationModule,
    ):
        self.communication_protocol = communication_protocol
        self.pin_r = PWM(Pin(pin_r))
        self.pin_g = PWM(Pin(pin_g))
        self.pin_b = PWM(Pin(pin_b))
        self.pin_fluo = Pin(pin_fluo, Pin.OUT)
        self.r_value = 0
        self.g_value = 0
        self.b_value = 0
        self.led_value = False
        self.fluo_value = False

    async def start(self):
        while True:
            message: DeviceMessage | None = self.communication_protocol.get_message()
            if not message:
                await asyncio.sleep(1)
                continue
            method_name = self.communication_protocol.extract_method_name_from_message(
                message
            )
            method = getattr(self, method_name)
            response = method(message)
            self.communication_protocol.send_message(response)

    def _update_led(self):
        self.pin_r.duty(self.r_value)
        self.pin_g.duty(self.g_value)
        self.pin_b.duty(self.b_value)

    def _turn_off_led(self):
        self.pin_r.duty(0)
        self.pin_g.duty(0)
        self.pin_b.duty(0)

    def _update_fluo(self, value):
        self.pin_fluo.value(value)

    ############################# REQUEST #############################

    def _set_settings_request(self, message: DeviceMessage) -> DeviceMessage:
        return self._set_settings_response(message)

    ############################# RESPONSE #############################
    def _set_settings_response(self, message: DeviceMessage) -> DeviceMessage:
        self.r_value = message.payload.get("color_r", self.r_value)
        self.g_value = message.payload.get("color_g", self.g_value)
        self.b_value = message.payload.get("color_b", self.b_value)
        self.led_value = message.payload.get("led_mode", self.led_value)
        self.fluo_value = message.payload.get("fluo_mode", self.fluo_value)
        self._update_led() if self.led_value else self._turn_off_led()
        self._update_fluo(self.fluo_value)
        return accept_message(message)

File: arduino/aquarium_esp8266_12E/src/aquarium.cpp
#include <Arduino.h>
#include <BasicMessage.h>
#include "aquarium.h"

Aquarium::Aquarium(ConfigManager& configManager, Mqtt& mqtt)
  : configManager(configManager), mqtt(mqtt){
  rPin = configManager.get("device.rPin").as<int>();
  gPin = configManager.get("device.gPin").as<int>();
  bPin = configManager.get("device.bPin").as<int>();
  fluoPin = configManager.get("device.fluoPin").as<int>();
  rValue = configManager.get("device.rValue").as<int>();
  gValue = configManager.get("device.gValue").as<int>();
  bValue = configManager.get("device.bValue").as<int>();
  ledValue = configManager.get("device.ledValue").as<bool>(); 
  fluoValue = configManager.get("device.fluoValue").as<bool>();

  pinMode(rPin, OUTPUT);
  pinMode(gPin, OUTPUT);
  pinMode(bPin, OUTPUT);
  pinMode(fluoPin, OUTPUT);
}

void Aquarium::onMessage(Message message){
  if(message.message_event == "set_settings"){
    setSettings(message);
    mqtt.sendMessage(basicResponse(message));
  }else if(message.message_event == "get_settings"){
    setSettings(message);
  }
}

void Aquarium::setSettings(Message message) {
    if (message.payload["color_r"].is<int>()) {
        rValue = message.payload["color_r"].as<int>();
        configManager.set("device.rValue",rValue);
    }
    if (message.payload["color_g"].is<int>()) {
        gValue = message.payload["color_g"].as<int>();
        configManager.set("device.gValue",rValue);
    }
    if (message.payload["color_b"].is<int>()) {
        bValue = message.payload["color_b"].as<int>();
        configManager.set("device.bValue",rValue);
    }
    if (message.payload["fluo_mode"].is<bool>()) {
        fluoValue = message.payload["fluo_mode"].as<bool>();
        configManager.set("device.fluoValue",fluoValue);
    }
    if (message.payload["led_mode"].is<bool>()) {
        ledValue = message.payload["led_mode"].as<bool>();
        configManager.set("device.ledValue",ledValue);
    }
    configManager.save();
    updateLed();
    updateFluo();
}

void Aquarium::updateLed() {
  if (ledValue) {
    analogWrite(rPin, rValue);
    analogWrite(gPin, gValue);
    analogWrite(bPin, bValue);
  } else {
    digitalWrite(rPin, LOW);
    digitalWrite(gPin, LOW);
    digitalWrite(bPin, LOW);
  }
}

void Aquarium::updateFluo() {
  if (fluoValue) {
    digitalWrite(fluoPin, HIGH);
  } else {
    digitalWrite(fluoPin, LOW);
  }
}